symengine/polys/upolybase.h

```diff
@@ -280,13 +330,121 @@ class UIntPolyBase : public UPolyBase<Container, Poly>
     {
     }
 
+    // returns degree of the poly
     inline unsigned int get_degree() const
     {
         return this->poly_.degree();
     }
-
+    // return coefficient of degree 'i'
     virtual integer_class get_coeff(unsigned int i) const = 0;
+    // return value of poly when ealudated at `x`
     virtual integer_class eval(const integer_class &x) const = 0;
+    // return `degree` + 1. `0` returned for zero poly.
+    virtual unsigned int size() const = 0;
+
+    RCP<const Basic> as_symbolic() const
+    {
+        auto it = (static_cast<const Poly &>(*this)).begin();
+        auto end = (static_cast<const Poly &>(*this)).end();
+
+        vec_basic args;
+        for (; it != end; ++it) {
+            integer_class m = to_integer_class(it->second);
+
+            if (it->first == 0) {
+                args.push_back(integer(m));
+            } else if (it->first == 1) {
+                if (m == 1) {
+                    args.push_back(this->var_);
+                } else {
+                    args.push_back(Mul::from_dict(integer(m),
+                                                  {{this->var_, one}}));
+                }
+            } else {
+                if (m == 1) {
+                    args.push_back(pow(this->var_, integer(it->first)));
+                } else {
+                    args.push_back(
+                        Mul::from_dict(integer(m),
+                                       {{this->var_, integer(it->first)}}));
+                }
+            }
+        }
+        return SymEngine::add(args);
+    }
+};
+
+template <typename T, typename Int>
+class ContainerBaseIter
+{
+protected:
+    RCP<const T> ptr_;
+    long i_;
+
+public:
+    ContainerBaseIter(RCP<const T> ptr, long x) : ptr_{ptr}, i_{x}
+    {
+    }
+
+    bool operator==(const ContainerBaseIter &rhs)
+    {
+        return (ptr_ == rhs.ptr_) and (i_ == rhs.i_);
+    }
+
+    bool operator!=(const ContainerBaseIter &rhs)
+    {
+        return not(*this == rhs);
+    }
+
+    std::pair<long, Int> operator*()
+    {
+        return std::make_pair(i_, ptr_->get_coeff_ref(i_));
+    }
+
+    std::shared_ptr<std::pair<long, Int>> operator->()
+    {
+        return std::make_shared<std::pair<long, Int>>(i_, ptr_->get_coeff_ref(i_));
+    }
+};
+
+template <typename T, typename Int>
+class ContainerForIter : public ContainerBaseIter<T, Int>
+{
+public:
+    ContainerForIter(RCP<const T> ptr, long x) : ContainerBaseIter<T, Int>(ptr, x)
+    {
+    }
+
+    ContainerForIter operator++()
+    {
+        this->i_++;
+        while (this->i_ < this->ptr_->size()) {
+            if (this->ptr_->get_coeff_ref(this->i_) != 0)
+                break;
+            this->i_++;
+        }
+        return *this;
+    }
```

Comment:
If we do that, then the reference has to be a const reference, because changing the global variable doesn't affect the polynomial.
But, if we throw an error, users will be less surprised with the behaviour